ORC-1906: Support `Meson` build

WillAyd opened the pull request:
### What changes were proposed in this pull request?

This adds a Meson configuration for the default C++ options in ORC. By adding this configuration, users of the Meson build system will have an easier time including Orc into their projects. In the mid to longer term, this could help the Apache Arrow project also become easier to include in other C++ project


### Why are the changes needed?
The current motivating factor is for Apache Arrow, where incremental improvements to the Meson build system are underway

### How was this patch tested?
A new CI job for Meson has been configured

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #2244

WillAyd commented:
closes #2244

wgtmac commented on subprojects/orc-format.wrap:
Perhaps it worths adding a README for this?

WillAyd replied:
Sounds good - happy to document wherever. Are you looking for just a quick blurb in subprojects/README, or do you think its best to add more comprehensive docs on using Meson in the main README (or elsewhere)?

dongjoon-hyun commented on subprojects/gtest.wrap:
```diff
@@ -0,0 +1,33 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
BTW, the directory name, `subprojects`, looks a little weird to me. Do you think we can have more meaningful and specific-to-Meson directory name instead of this general term `subprojects`?

WillAyd replied:
You do have the option to override this but Meson really discourages it. It's a convention across all projects, and there's a higher risk of introducing bugs from overriding that default

dongjoon-hyun replied:
Got it.

dongjoon-hyun commented on meson.build:
```diff
@@ -0,0 +1,43 @@
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
```
It seems that this is inconsistent. Could you follow the header style from the other files?

dongjoon-hyun commented on subprojects/google-snappy.wrap:
```diff
@@ -0,0 +1,29 @@
+# under the License.
+
+[wrap-file]
+directory = snappy-1.1.9
```
Please use Snappy 1.2.2.

https://github.com/apache/orc/blob/b6841129e0bc81c86276a7d6f19b497b20dc3bd3/cmake_modules/ThirdpartyToolchain.cmake#L24

dongjoon-hyun commented on subprojects/gtest.wrap:
```diff
@@ -0,0 +1,33 @@
+# under the License.
+
+[wrap-file]
+directory = googletest-1.17.0
```
If this is a new version, we need to upgrade `CMake` first before merging this `Meson` PR.

https://github.com/apache/orc/blob/b6841129e0bc81c86276a7d6f19b497b20dc3bd3/cmake_modules/ThirdpartyToolchain.cmake#L26

dongjoon-hyun commented on subprojects/lz4.wrap:
```diff
@@ -0,0 +1,30 @@
+# under the License.
+
+[wrap-file]
+directory = lz4-1.10.0
```
This is identical.

dongjoon-hyun commented on subprojects/protobuf.wrap:
```diff
@@ -0,0 +1,33 @@
+# under the License.
+
+[wrap-file]
+directory = protobuf-25.2
```
This looks tricky because C++ module is still using old one, 3.5.1.

https://github.com/apache/orc/blob/b6841129e0bc81c86276a7d6f19b497b20dc3bd3/cmake_modules/ThirdpartyToolchain.cmake#L27

WillAyd commented:
Thanks @eli-schwartz that context is really helpful.

@dongjoon-hyun for now I have downgrade the versions pulled in by Meson's wrapdb for googletest and protobuf to match the existing CMake configuration. The snappy version is the only one left I think that is different, but that specific version doesn't exist in the wrapdb yet. I'll take a look at adding that

WillAyd commented:
> The snappy version is the only one left I think that is different, but that specific version doesn't exist in the wrapdb yet. I'll take a look at adding that

google-snappy 1.2.2 was added upstream to the Meson WrapDB today and I've updated this PR to use that. Hope I've addressed all your feedback @dongjoon-hyun but let me know if anything is missing

dongjoon-hyun commented on c++/src/meson.build:
```diff
@@ -0,0 +1,196 @@
+)
+
+source_files = [adaptor_header]
+source_files += files(
```
Just a question. Do we need to maintain this source file list manually? It looks like a little error prone.

eli-schwartz replied:
You already maintain it manually in CMakeLists.txt, it would need to be updated both places at the same time (and checked in CI).

But it's possible to maintain it once in a text file. For meson, that looks like this:

```meson
source_files += files(
    import('fs').read('all_sources.lst').split()
)
```
with one source file per line in `all_sources.lst`. And similar logic for cmake.

eli-schwartz commented on README.md:
````diff
@@ -115,3 +115,43 @@ Cmake option BUILD_ENABLE_AVX512 can be set to "ON" or (default value)"OFF" at t
+While CMake is the official build system for orc, there is unofficial support for using Meson to build select parts of the project. To build a debug version of the library and test it using Meson, from the project root you can run:
+
+```shell
+meson setup builddir
````
> Otherwise, we need to add `builddir` to `.gitignore` additionally. So, we prefer (1).

You do not! :) Meson seeks to provide such quality of life features automatically. Besides, the README is irrelevant, users will do whatever they want and expect it to work. Try to handle `build-release/` and `build-debug/`. :)

https://mesonbuild.com/FAQ.html#how-do-i-ignore-the-build-directory-in-my-vcs

You must have used `mkdir` rather than `meson setup` to test...

dongjoon-hyun commented on c++/test/meson.build:
```diff
@@ -0,0 +1,85 @@
+    '../src',
+)
+
+test_sources = [
```
ditto. How did you generate this list? Do we need to maintain this manually?

dongjoon-hyun commented on meson.build:
```diff
@@ -0,0 +1,43 @@
+    'cpp',
+    version: '2.2.0-SNAPSHOT',
+    license: 'Apache-2.0',
+    meson_version: '>=1.3.0',
```
Given that this is for Apache ORC 2.2.0 (Year 2026), can we start with `1.8.0` or `1.8.2`?

eli-schwartz replied:
Perhaps, but it's more convenient to not artificially restrict what works.

Also, Debian Trixie will not have Meson 1.8.* because it was released after Trixie entered release freeze. So it will be quite some time before a stable version of Debian could build this, unless the user manually installs Meson as well.

WillAyd replied:
In addition to what @eli-schwartz has already said, this is the same version constraint used by Apache Arrow, Apache nanoarrow, and Apache arrow-adbc projects, so I think its great to keep consistent across the projects.

dongjoon-hyun commented on subprojects/packagefiles/orc-format/meson.build:
```diff
@@ -0,0 +1,57 @@
+    'cpp',
+    version: '1.1.0',
+    license: 'Apache-2.0',
+    meson_version: '>=1.3.0',
```
ditto. Can we start with 1.8.2 (or 1.8.0)?

dongjoon-hyun commented on subprojects/protobuf.wrap:
```diff
@@ -0,0 +1,25 @@
+# under the License.
+
+[wrap-file]
+directory = protobuf-3.5.1
```
Thank you.

dongjoon-hyun commented on .github/workflows/build_and_test.yml:
```diff
@@ -248,3 +248,17 @@ jobs:
 
+  meson:
+    name: "Meson C++ configuration"
+    runs-on: ubuntu-latest
```
1. To make it sure the supported OSes, please use OS matrix like the existing system at least.

https://github.com/apache/orc/blob/63541a231b9c7317c755b675dd3ec10541735c0d/.github/workflows/build_and_test.yml#L62-L72

2. In addition to the above OSes, CI has the following test coverage too. However, I believe (1) is enough for `Meson`.

https://github.com/apache/orc/blob/63541a231b9c7317c755b675dd3ec10541735c0d/.github/workflows/build_and_test.yml#L40-L51

WillAyd commented on tools/meson.build:
```diff
@@ -0,0 +1,34 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# 
+#     http://www.apache.org/licenses/LICENSE-2.0
```
Ah sorry - nice catch. Fixed!

dongjoon-hyun reviewed (dismissed):
Thank you for update, @WillAyd . However, this seems to be ignored from my perspective.

https://github.com/apache/orc/pull/2249/files#r2136861282

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you so much, @WillAyd and @eli-schwartz .

dongjoon-hyun commented:
cc @williamhyun , @wgtmac , @ffacs , too.

williamhyun reviewed (approved):
+1 LGTM, will be merging this now.

williamhyun commented:
Thank you @WillAyd, @eli-schwartz, and @dongjoon-hyun for your contributions!

WillAyd commented:
@williamhyun you can use william.ayd@icloud.com - thanks!

dongjoon-hyun commented:
> you can use [william.ayd@icloud.com](mailto:william.ayd@icloud.com) - thanks!

FYI, JIRA ID means the ID at https://issues.apache.org , @WillAyd .

dongjoon-hyun commented:
No problem at all. I added you to the Apache ORC contributor group and assigned ORC-1906 to you.
